openssl/src/ts.rs

```diff
@@ -0,0 +1,501 @@
+use std::ptr;
+
+use crate::asn1::{Asn1IntegerRef, Asn1ObjectRef, Asn1OctetString, Asn1StringRef};
+use crate::bio::MemBioSlice;
+use crate::error::ErrorStack;
+use crate::hash::{Hasher, MessageDigest};
+use crate::pkey::{HasPrivate, PKeyRef};
+use crate::util::{ForeignTypeExt, ForeignTypeRefExt};
+use crate::x509::{X509Algorithm, X509AlgorithmRef, X509Ref};
+use crate::{cvt, cvt_p};
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_MSG_IMPRINT;
+    fn drop = ffi::TS_MSG_IMPRINT_free;
+
+    /// A message imprint contains the hash of the data to be timestamped.
+    pub struct TsMsgImprint;
+
+    /// Reference to `TsMsgImprint`.
+    pub struct TsMsgImprintRef;
+}
+
+impl PartialEq for TsMsgImprintRef {
+    fn eq(&self, other: &Self) -> bool {
+        self.get_msg() == other.get_msg() && self.get_algo() == other.get_algo()
+    }
+}
+impl PartialEq for TsMsgImprint {
+    fn eq(&self, other: &Self) -> bool {
+        TsMsgImprintRef::eq(self, other)
+    }
+}
+
+impl TsMsgImprint {
+    /// Creates a new message imprint.
+    #[corresponds(TS_MSG_IMPRINT_new)]
+    pub fn new() -> Result<TsMsgImprint, ErrorStack> {
+        ffi::init();
+        unsafe {
+            let imprint = cvt_p(ffi::TS_MSG_IMPRINT_new())?;
+            Ok(TsMsgImprint::from_ptr(imprint))
+        }
+    }
+
+    /// Sets the algorithm identifier of the message digest algorithm.
+    #[corresponds(TS_MSG_IMPRINT_set_algo)]
+    pub fn set_algo(&mut self, algo: &X509AlgorithmRef) -> Result<(), ErrorStack> {
+        unsafe { cvt(ffi::TS_MSG_IMPRINT_set_algo(self.as_ptr(), algo.as_ptr())).map(|_| ()) }
+    }
+
+    /// Sets the message **digest** of the data to be timestamped.
+    /// It is named this way to match the name in openssl itself
+    #[corresponds(TS_MSG_IMPRINT_set_msg)]
+    pub fn set_msg(&mut self, digest: &[u8]) -> Result<(), ErrorStack> {
+        let len = if digest.len() > c_int::MAX as usize {
+            panic!("digest length is too large");
+        } else {
+            digest.len() as c_int
+        };
+
+        unsafe {
+            cvt(ffi::TS_MSG_IMPRINT_set_msg(
+                self.as_ptr(),
+                digest.as_ptr() as *mut _,
+                len,
+            ))
+            .map(|_| ())
+        }
+    }
+
+    /// Creates a ready-to-use message imprint from a message and a specified hash algorithm.
+    pub fn from_message_with_algo(msg: &[u8], md: MessageDigest) -> Result<Self, ErrorStack> {
+        let mut h = Hasher::new(md)?;
+        h.update(msg)?;
+        let hash = h.finish()?;
+        Self::from_prehash_with_algo(&hash, md)
+    }
+
+    /// Creates a ready-to-use message imprint from the hash of a message and a specified hash algorithm.
+    ///
+    /// `hash` must have originated from the hash function specified by `md`.
+    pub fn from_prehash_with_algo(hash: &[u8], md: MessageDigest) -> Result<Self, ErrorStack> {
+        let mut algo = X509Algorithm::new()?;
+        algo.set_md(md);
```

Comment:
By using `set_md()` that wraps `X509_ALGOR_set_md()`, you will have a different encoding of the `AlgorithmIdentifier`'s parameters than OpenSSL uses in their [`ts.c`][1]. This PR's code will omit the parameters for the SHA-1/SHA-2 families of hashes, while OpenSSL encodes them as an explicit ASN.1 NULL for all `md`. While your choice is likely the correct one -- the backstory is an [incorrect translation][2] from older to newer ASN.1 syntax -- it poses an interop concern with OpenSSL.

[1]: https://github.com/openssl/openssl/blob/d1c2c054a4b585eed8c883367d80e2a972c4846f/apps/ts.c#L464-L470
[2]: https://www.rfc-editor.org/rfc/rfc5754#page-4

Reply:
what's the best way forward here?

should I special case OpenSSL's X509_ALGOR to be a struct exposing the underlying fields such that I can exactly replicate the behaviour in ts.c?

Reply:
I think you can simply expose and use `X509_ALGOR_set0`.